### ISSUE TITLE
Bal build fails with a BAD SAD

### ISSUE THREAD

**Issue by hasithaa** [ballerina-platform/ballerina-lang#30705]
Observed following with `bal build` command for  https://github.com/jclark/semtype.  

Pack: SLBeta.1-RC3

```
[2021-05-21 09:54:32,013] SEVERE {b7a.log.crash} - class io.ballerina.compiler.api.impl.symbols.BallerinaTypeReferenceTypeSymbol cannot be cast to class io.ballerina.compiler.api.symbols.FunctionTypeSymbol (io.ballerina.compiler.api.impl.symbols.BallerinaTypeReferenceTypeSymbol and io.ballerina.compiler.api.symbols.FunctionTypeSymbol are in unnamed module of loader 'app') 
java.lang.ClassCastException: class io.ballerina.compiler.api.impl.symbols.BallerinaTypeReferenceTypeSymbol cannot be cast to class io.ballerina.compiler.api.symbols.FunctionTypeSymbol (io.ballerina.compiler.api.impl.symbols.BallerinaTypeReferenceTypeSymbol and io.ballerina.compiler.api.symbols.FunctionTypeSymbol are in unnamed module of loader 'app')
	at io.ballerina.compiler.api.impl.SymbolFactory.createFunctionSymbol(SymbolFactory.java:251)
	at io.ballerina.compiler.api.impl.SymbolFactory.getBCompiledSymbol(SymbolFactory.java:138)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.lookupSymbol(BallerinaSemanticModel.java:281)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.symbol(BallerinaSemanticModel.java:137)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:157)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:68)
	at io.ballerina.compiler.syntax.tree.NodeTransformer.transform(NodeTransformer.java:192)
	at io.ballerina.compiler.syntax.tree.RequiredParameterNode.apply(RequiredParameterNode.java:55)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.apply(SyntaxTreeMapGenerator.java:368)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:87)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:68)
	at io.ballerina.compiler.syntax.tree.NodeTransformer.transform(NodeTransformer.java:476)
	at io.ballerina.compiler.syntax.tree.FunctionSignatureNode.apply(FunctionSignatureNode.java:59)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.apply(SyntaxTreeMapGenerator.java:368)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:91)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:68)
	at io.ballerina.compiler.syntax.tree.NodeTransformer.transform(NodeTransformer.java:48)
	at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.apply(SyntaxTreeMapGenerator.java:368)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:87)
	at org.ballerinalang.diagramutil.SyntaxTreeMapGenerator.transformSyntaxNode(SyntaxTreeMapGenerator.java:68)
	at io.ballerina.compiler.syntax.tree.NodeTransformer.transform(NodeTransformer.java:44)
	at org.ballerinalang.diagramutil.DiagramUtil.getSyntaxTreeJSON(DiagramUtil.java:50)
	at org.ballerinalang.observability.anaylze.DefaultObservabilitySymbolCollector.process(DefaultObservabilitySymbolCollector.java:97)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:135)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:114)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:152)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:113)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:66)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:302)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
 ```

**Comment by pubudu91** [comment]
It seems like the above issue is due to the changes introduced in #30256. Affects the following scenario:
```ballerina
type BinOp function (int) returns string;

type Foo readonly & record {|
    BinOp <cursor>union;
|};
```
For cases like that, although it's a field/variable, we create a function symbol. With the changes in the above PR, when creating the type for the `union` field, it returns a type ref type symbol. Will change the behaviour for the above case to take the typedesc itself, instead of the type ref.

**Comment by pubudu91** [comment]
Fixed with #30713
